BraintreeCore/src/main/java/com/braintreepayments/api/core/ConfigurationLoader.kt

```diff
@@ -2,58 +2,89 @@ package com.braintreepayments.api.core
 ) {
     constructor(context: Context, httpClient: BraintreeHttpClient) : this(
         httpClient, ConfigurationCache.getInstance(context)
     )
 
     fun loadConfiguration(authorization: Authorization, callback: ConfigurationLoaderCallback) {
         if (authorization is InvalidAuthorization) {
-            val message = authorization.errorMessage
-            // NOTE: timing information is null when configuration comes from cache
-            callback.onResult(null, BraintreeException(message), null)
+            val invalidAuthError = BraintreeException(authorization.errorMessage)
+            callback.onResult(ConfigurationLoaderResponse(error = invalidAuthError))
             return
         }
-        val configUrl = Uri.parse(authorization.configUrl)
+        loadConfigurationInBackground(authorization, callback)
+    }
+
+    private fun loadConfigurationInBackground(
+        auth: Authorization,
+        callback: ConfigurationLoaderCallback
+    ) {
+        val url = Uri.parse(auth.configUrl)
             .buildUpon()
             .appendQueryParameter("configVersion", "3")
             .build()
             .toString()
 
-        val cachedConfig = configurationCache.getConfiguration(authorization, configUrl)
-        cachedConfig?.let {
-            callback.onResult(cachedConfig, null, null)
-        } ?: run {
-            val request = InternalHttpRequest(method = HttpMethod.GET, path = configUrl)
-            httpClient.sendRequest(
-                request = request,
-                authorization = authorization
-            ) { response, httpError ->
-                val responseBody = response?.body
-                val timing = response?.timing
-                if (responseBody != null) {
-                    try {
-                        val configuration = Configuration.fromJson(responseBody)
-                        configurationCache.putConfiguration(configuration, authorization, configUrl)
-                        callback.onResult(configuration, null, timing)
-                    } catch (jsonException: JSONException) {
-                        callback.onResult(null, jsonException, null)
-                    }
-                } else {
-                    httpError?.let { error ->
-                        val errorMessageFormat = "Request for configuration has failed: %s"
-                        val errorMessage = String.format(errorMessageFormat, error.message)
-                        val configurationException = ConfigurationException(errorMessage, error)
-                        callback.onResult(null, configurationException, null)
-                    }
-                }
+        val cbRef = WeakReference(callback)
+        threadScheduler.runOnBackground {
+            val response = loadConfigFromCache(url, auth) ?: loadConfigFromNetwork(url, auth)
+            threadScheduler.runOnMain {
+                val cb = cbRef.get()
```

Comment:
Can we update these names to be more explicit? It took me a bit to grock that they stood for `callbackReference` and `callback`. Also not sure if the second can just be reduced to `callbackReference.get().onResult(response)`?

Reply:
Yeah that's a good call. Just updated here: [4fcddf5](https://github.com/braintree/braintree_android/pull/1133/commits/4fcddf5a6828c7b4d502a14791c530e586a8cfee)